backtracking/all-combinations-of-size-k.ts

```diff
@@ -1,11 +1,22 @@
+/**
+ * This generates an array of unique sub"sets" (represented by ascendingly sorted subarrays)
+ * of size k out of n+1 numbers from 1 to n.
+ *
+ * By using a backtracking algorithm we can incrementally build sub"sets" while dropping candidates
+ * that cannot contribute anymore to a valid solution.
+ * Steps:
+ * - From the starting number (i.e. "1") generate all combinations of k numbers.
+ * - Once we got all combinations for the given number we can discard it (“backtracks”)
+ *   and repeat the same process for the next number.
+ */
 export function generateCombinations(n: number, k: number): number[][] {
-  return generateAllCombos(n, k);
+  return generateAllCombos(n, k, 1);
 }
 
 function generateAllCombos(
   n: number,
   k: number,
-  startCursor: number = 1,
+  startCursor: number,
   combinationsAcc: number[][] = [],
```

Comment:
Why are you keeping those as default arguments?

Reply:
Those only make sense for the sake of recursive calls. Do you see any advantage removing default and providing starting values from the main function?

Reply:
Two advantages:

1. Conceptually, these aren't really default arguments; these are starting arguments.
2. It *may* be negligibly more efficient: Default arguments are typically implemented as "if it wasn't passed, then set the default value". This requires checking whether these variables were passed.

IMO it would probably be cleanest to have this private "generate all combos" utility be a closure receiving `combinationsAcc` and `currentCombination` as "upvalues":

```javascript
export function generateCombinations(n: number, k: number): number[][] {
  let combinationsAcc = [];
  let currentCombination = [];
  const generateAllCombos = (n, k, startCursor) => { ... };
  return generateAllCombos(n, k, 1);
}
```

Reply:
Good points. 

About performance, we're talking about _micro optimizations_ that could fluctuate across different runtimes. Btw as you said there shouldn't be a tangible difference between the two implementations from a performance POV